docs/docs/reference/gen_notebooks/dspy_prompt_optimization.md

```diff
@@ -0,0 +1,290 @@
+## Enable Tracking using Weave
+
+Weave is currently integrated with DSPy, and including [`weave.init`](../docs/reference/python-sdk/weave/index.md) at the start of our code lets us automatically trace our DSPy functions which can be explored in the Weave UI. Check out the [Weave integration docs for DSPy](../docs/guides/integrations/dspy.md) to learn more.
+
+
+```python
+import weave
+
+weave.init(project_name="dspy-bigbench-hard")
+```
+
+In this tutorial, we use a metadata class inherited from [`weave.Model`](../docs/guides/core-types/models.md) to manage our metadata.
+
+
+```python
+class Metadata(weave.Model):
+    dataset_address: str = "maveriq/bigbenchhard"
+    big_bench_hard_task: str = "causal_judgement"
+    num_train_examples: int = 50
+    openai_model: str = "gpt-3.5-turbo"
+    openai_max_tokens: int = 2048
+    max_bootstrapped_demos: int = 8
+    max_labeled_demos: int = 8
+
+
+metadata = Metadata()
+```
+
+| ![](../static/img/dspy_prompt_optimiztion/metadata.gif) |
+|---|
+| The `Metadata` objects are automatically versioned and traced when functions consuming them are traced |
+
+## Load the BIG-Bench Hard Dataset
+
+We will load this dataset from HuggingFace Hub, split into training and validation sets, and [publish](../docs/guides/core-types/datasets.md) them on Weave, this will let us version the datasets, and also use [`weave.Evaluation`](../docs/guides/core-types/evaluations.md) to evaluate our prompting strategy.
+
+
+```python
+import dspy
+from datasets import load_dataset
+
+
+@weave.op()
+def get_dataset(metadata: Metadata):
+    # load the BIG-Bench Hard dataset corresponding to the task from Huggingface Hug
+    dataset = load_dataset(metadata.dataset_address, metadata.big_bench_hard_task)[
+        "train"
+    ]
+
+    # create the training and validation datasets
+    rows = [{"question": data["input"], "answer": data["target"]} for data in dataset]
+    train_rows = rows[0 : metadata.num_train_examples]
+    val_rows = rows[metadata.num_train_examples :]
+
+    # create the training and validation examples consisting of `dspy.Example` objects
+    dspy_train_examples = [
+        dspy.Example(row).with_inputs("question") for row in train_rows
+    ]
+    dspy_val_examples = [dspy.Example(row).with_inputs("question") for row in val_rows]
+
+    # publish the datasets to the Weave, this would let us version the data and use for evaluation
+    weave.publish(
+        weave.Dataset(
+            name=f"bigbenchhard_{metadata.big_bench_hard_task}_train", rows=train_rows
+        )
+    )
+    weave.publish(
+        weave.Dataset(
+            name=f"bigbenchhard_{metadata.big_bench_hard_task}_val", rows=val_rows
+        )
+    )
+
+    return dspy_train_examples, dspy_val_examples
+
+
+dspy_train_examples, dspy_val_examples = get_dataset(metadata)
+```
+
+| ![](../static/img/dspy_prompt_optimiztion/datasets.gif) |
+|---|
+| The datasets, once published, can be explored in the Weave UI |
+
+## The DSPy Program
```

Comment:
nit: Program sounds weird here to me for some reason. Is this an official term DSPy uses?

Reply:
Yes, `program` is the official term that is used by DSPy.